libethash-cuda/CUDAMiner.cpp

```diff
@@ -106,18 +106,16 @@ CUDAMiner::~CUDAMiner()
 	delete m_hook;
 }
 
-void CUDAMiner::report(uint64_t _nonce)
+void CUDAMiner::report(uint64_t _nonce, const WorkPackage& w)
 {
 	// FIXME: This code is exactly the same as in EthashGPUMiner.
-	WorkPackage w = work();  // Copy work package to avoid repeated mutex lock.
 	Result r = EthashAux::eval(w.seed, w.header, _nonce);
```

Comment:
yes, yes, yes... much better. I like this much better and will withdraw my solution. Wish there was a way to get rid of the miner x_work lock though. I thought that by only calling work at the top of the work loop the race window would be small enough to ignore, but apparently not.

Reply:
There must be some sync between stratum and miners threads. We can use atomics probably, but that's more tricky.